src/ILVerify/src/Verifier.cs

```diff
@@ -0,0 +1,224 @@
+            }
+            catch (BadImageFormatException)
+            {
+                output.AppendLine("Unable to resolve token");
+                numErrors++;
+            }
+            catch (PlatformNotSupportedException e)
+            {
+                output.AppendLine(e.Message);
+                numErrors++;
+            }
+            catch (VerifierException e)
+            {
+                output.AppendLine(e.Message);
+                numErrors++;
+            }
+            catch (TypeSystemException e)
+            {
+                output.AppendLine(e.Message);
+                numErrors++;
+            }
+
+            return new VerificationResult() { NumErrors = numErrors, Message = output.ToString(), _errors = errors };
+        }
+
+        internal void AppendError(MethodDesc method, string moduleName, VerificationErrorArgs args, StringBuilder output)
+        {
+            output.Append("[IL]: Error: ");
+
+            output.Append("[");
+            output.Append(moduleName);
+            output.Append(" : ");
+            output.Append(((EcmaType)method.OwningType).Name);
+            output.Append("::");
+            output.Append(method.Name);
+            output.Append("(");
+            if (method.Signature._parameters != null && method.Signature._parameters.Length > 0)
+            {
+                foreach (TypeDesc parameter in method.Signature._parameters)
+                {
+                    output.Append(parameter.ToString());
+                    output.Append(", ");
+                }
+                output.Remove(output.Length - 2, 2);
+            }
+            output.Append(")");
+            output.Append("]");
+
+            output.Append("[offset 0x");
+            output.Append(args.Offset.ToString("X8"));
+            output.Append("]");
+
+            if (args.Found != null)
+            {
+                output.Append("[found ");
+                output.Append(args.Found);
+                output.Append("]");
+            }
+
+            if (args.Expected != null)
+            {
+                output.Append("[expected ");
+                output.Append(args.Expected);
+                output.Append("]");
+            }
+
+            if (args.Token != 0)
+            {
+                output.Append("[token  0x");
+                output.Append(args.Token.ToString("X8"));
+                output.Append("]");
+            }
+
+            output.Append(" ");
+            var str = _stringResourceManager.Value.GetString(args.Code.ToString(), CultureInfo.InvariantCulture);
+            output.AppendLine(string.IsNullOrEmpty(str) ? args.Code.ToString() : str);
+        }
+    }
+
+    public class VerificationResult
+    {
+        public int NumErrors = 0;
+        public string Message = string.Empty;
+        internal IEnumerable<VerifierError> _errors; // Note: there may be fewer errors recorded here than counted in NumErrors, which also counts exceptions
+    }
```

Comment:
I agree with @jkotas. I think it would make sense to simply have an enumeration of `VerificationError`s (which can be an array internally).
I think a new `VerificationError` class would be helpful for this, which could look like this:
```csharp
class VerificationError
{
    public readonly string ModuleName;
    public readonly string MethodName;
    public readonly VerificationErrorArgs ErrorArgs;
}
```
It might even make sense to use a `MethodDefinitionHandle` and `PEReader` instead of strings, to be more consistent with the proposed `Verify` method.
I am currently working on a use case, where having the VerificationErrorArgs for each reported error is vital. Creating the error string can then also be implemented by the application using the Verifier rather than the Verifier having to implement it.

Reply:
I went with strings. Feel free to adjust to your need, after this PR but before we publish.